File: 2024/08/09/133. Clone Graph/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (!node) return nullptr;
        unordered_map<Node*, Node*> res;
        return dfs(node,res);
    }

    Node* dfs(Node* node, unordered_map<Node*, Node*>& res) {
        if (res.find(node) != res.end()) {
            return res[node];
        }
        Node* copy = new Node(node->val);
        res[node]=copy;
        for (Node* neighbor : node->neighbors) {
            copy->neighbors.push_back(dfs(neighbor, res));
        }
        return copy;
    }

};

File: 2024/08/09/133. Clone Graph/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        def helper(node, hmap):
            mynode = Node(node.val)
            hmap[node] = mynode
            for n in node.neighbors:
                if n not in hmap:
                    helper(n, hmap)
                mynode.neighbors.append(hmap[n])
        if not node: return None
        hmap = {}
        helper(node, hmap)
        return hmap[node]